packages/vhd-lib/merge.js

```diff
@@ -79,30 +89,44 @@ module.exports = limitConcurrency(2)(async function merge(
     }
 
     // counts number of allocated blocks
-    let nBlocks = 0
+    const toMerge = []
     for (let block = mergeState.currentBlock; block < maxTableEntries; block++) {
       if (childVhd.containsBlock(block)) {
-        nBlocks += 1
+        toMerge.push(block)
       }
     }
-
+    const nBlocks = toMerge.length
     onProgress({ total: nBlocks, done: 0 })
 
-    // merges blocks
-    for (let i = 0; i < nBlocks; ++i, ++mergeState.currentBlock) {
-      while (!childVhd.containsBlock(mergeState.currentBlock)) {
-        ++mergeState.currentBlock
-      }
-
-      await parentHandler.writeFile(mergeStatePath, JSON.stringify(mergeState), { flags: 'w' }).catch(warn)
+    const merging = new Set()
+    let counter = 0
 
-      mergeState.mergedDataSize += await parentVhd.coalesceBlock(childVhd, mergeState.currentBlock)
-      onProgress({
-        total: nBlocks,
-        done: i + 1,
-      })
-    }
+    const mergeStateWriter = makeThrottledWriter(parentHandler, mergeStatePath, 10e3)
 
+    await asyncEach(
+      toMerge,
+      async blockId => {
+        merging.add(blockId)
+        try {
+          mergeState.mergedDataSize += await parentVhd.coalesceBlock(childVhd, blockId)
+        } catch (e) {
+          // @todo what should I do ?
```

Comment:
Should the merge continue in that case?

Reply:
I think, we should not continue merging, it may lead to corrupted data, but what should we do ? log an error ? delete the potentially broken vhd ? 

also the data may be safe , it can be a connexion problem or any other external factor

Reply:
Simply don't catch the error and `asyncEach` will forward it, the rest of the code should properly handle that.